### ISSUE TITLE
Too many call on Service Catalog ?

### ISSUE THREAD

**Issue by Menahem1** [aws-ia/terraform-aws-control_tower_account_factory#403]
**AFT Version:** 1.10.4

**Bug Description**
Too many API call on Service Catalog 
Actually i have more than 30.000 API Calls on Service Catalog, is this normal ?

On CloudTrail i see many recurrence
![Opera Instantané_2023-10-25_103151_eu-west-3 console aws amazon com](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/assets/1558463/85d60364-f484-472d-8fc2-ebec2daf9311)

Any idea why ?

Thanks

**Comment by PeterBengtson** [comment]
I've been using AFT for years, and I'm also worried about this. In a system
where no new accounts are created and no accounts updated, I still see lots
of Service Catalog calls, to the point where the free tier is exhausted
many times over.

AFT is obviously quite inefficient and expensive in this respect. I see no
reason why this incessant polling should take place. Perhaps its creators
can elucidate.

**Comment by hanafya** [comment]
Hey @Menahem1! Can you tell us the time range where you are seeing the API calls? In addition, how many customizations are you running?

**Comment by Menahem1** [comment]
Hey @hanafya 
No time range is configured on the screenshot (on the monthly bill i see more that 30.000 api calls on September)
Juste one customizations and very few accounts created

**Comment by hanafya** [comment]
Hey @Menahem1 How many accounts are in your AWS Organization?

**Comment by Menahem1** [comment]
about 15

**Comment by giedriuskilcauskas** [comment]
The guilty one is here: https://github.com/aws-ia/terraform-aws-control_tower_account_factory/blob/main/modules/aft-account-request-framework/eventbridge.tf#L68
That lambda runs every 5 mins and makes many calls to Service catalog API. As a quick workaround we just manually changed that to like 12hours (even that is too much). Ideally, it should be triggered by SQS only

**Comment by PeterBengtson** [comment]
Has this been fixed?

**Comment by robbycuenot** [comment]
I've noticed this as well. Removing the NAT gateway requirement dramatically reduced the cost of running AFT (from $100/month to $50/month, roughly). Now, the Service Catalog calls make up the bulk of the cost.

![image](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/assets/51327557/1cff168a-3dd6-4684-87d7-ea5cce7db4da)

If this could be changed in a new minor release, it would halve the cost again. The screenshot above is for an AFT setup with 8 accounts, and no new accounts created this calendar year.

**Comment by wellsiau-aws** [comment]
I suspect this part of the [aft_account_request_processor](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/blob/aee036db9d4926d60fd48a04fbc1d70917749456/src/aft_lambda/aft_account_request_framework/aft_account_request_processor.py#L43-L49) is causing excessive call.

it might be a good intention to throw warning earlier, but I argue this call should begin only when there's a new message in the SQS queue.

**Comment by PeterBengtson** [comment]
Any progress on this? Any reaction from the dev team? Any activity at all??

**Comment by Tharsan05** [comment]
Hi, Any update on this ?

Due to too many API calls within short time period, when there are lot of provisioned products exists (Total accounts 250+). 
```
we are getting below error as well.
An error occurred in the 'aft-account-request-processor' Lambda function.

Error Message: An error occurred (ThrottlingException) when calling the 
ScanProvisionedProducts operation (reached max retries: 4): Rate exceeded

```

**Comment by jpablos11** [comment]
Hi Everyone, 

Adding to this conversation, I was thinking on opening a New issue but I decided to do a search first and bumped into this one

Fresh new install, and saw Service Catalog cost rising up which did not make much sense at first, digging deeper, I found out exactly what is reported on this thread:

- The "aft-account-request-processor" lambda has an EventBridge "rate(5 minutes)" trigger, and
- The Lambda is calling Service Catalog every time, disregarding if there are new messages to process, which as mentioned above is inefficient and wasteful

A screenshot from Cost Explorer, Dimension = "Usage type", Service = "Service Catalog":
![image](https://github.com/user-attachments/assets/2d2f0eb5-5ae0-4039-a4eb-fe81a5afd61d)

There have been days where we have not used AFT for sure, so you can see an almost consistent charge of $0.81 cents per day, it is not much, but considering is not adding any value, I believe this is something that should be fixed

Here is an screenshot from the "aft-account-request-processor" lambda logs where seems to show that Service Catalog is called first, before actually verifying if there are any messages to process in SQS:
![image](https://github.com/user-attachments/assets/a66f57da-fbea-40d1-a42d-7ced88f662f3)

A better approach would be to either call Service Catalog ONLY IF there were messages in the queue (check SQS first), or to trigger the Lambda only based on events in SQS; either of the approaches should take care of the issue

**Comment by jpablos11** [comment]
FYI, I have just disabled manually the trigger on EventBridge for now, I will observe cost for a couple of days and report my findings:
![image](https://github.com/user-attachments/assets/79ef66e6-c0f4-4d2a-8298-daea3b3828a8)

**Comment by Tharsan05** [comment]
> Hi Everyone,
> 
> Adding to this conversation, I was thinking on opening a New issue but I decided to do a search first and bumped into this one
> 
> Fresh new install, and saw Service Catalog cost rising up which did not make much sense at first, digging deeper, I observed the same that is reported on this thread:
> 
> * The "aft-account-request-processor" lambda has an EventBridge "rate(5 minutes)" trigger, and
> * The Lambda is calling Service Catalog every time, disregarding if there were new messages to process, which I think is inefficient and wasteful
> 
> A screenshot from Cost Explorer, Dimension = "Usage type", Service = "Service Catalog": ![image](https://private-user-images.githubusercontent.com/1699773/385624222-2d2f0eb5-5ae0-4039-a4eb-fe81a5afd61d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0OTU3NjQsIm5iZiI6MTczMTQ5NTQ2NCwicGF0aCI6Ii8xNjk5NzczLzM4NTYyNDIyMi0yZDJmMGViNS01YWUwLTQwMzktYTRlYi1mZTgxYTVhZmQ2MWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTNUMTA1NzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDY4MzRkNzVhZjIwMmRmMmUzOGY2OTQwOWU0NWQwYjQ2NzY4NWZmNWNmMzM5YWZlOTU3MzE3MzdhMjI5YmZiOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zQWBJHFqvSe1YAS-WLugoHvoAXn_2KwDbnb43qlzttk)
> 
> There have been days where we have not used AFT for sure, so you can see an almost consistent charge of $0.81 cents per day, it is not much (it might become significant as the number of accounts grows), but considering is not adding any value, I believe this is something that should be fixed
> 
> Here is an screenshot from the "aft-account-request-processor" lambda logs where seems to show that Service Catalog is being called first and some info extracted, this before actually verifying if there were any messages to process in SQS: ![image](https://private-user-images.githubusercontent.com/1699773/385698408-5169c077-ce21-4d17-894b-b421efcf4782.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0OTU3NjQsIm5iZiI6MTczMTQ5NTQ2NCwicGF0aCI6Ii8xNjk5NzczLzM4NTY5ODQwOC01MTY5YzA3Ny1jZTIxLTRkMTctODk0Yi1iNDIxZWZjZjQ3ODIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTNUMTA1NzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTdkMzRhN2NhZWRiOWM5OWU1YWY5YzBkYmZlNmYxNmExYWZiNWE3M2ZmN2M5NGU3NGMyNmQ2MzZjMWNmMjI4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ViV8Ae77tkvXmjcojKFJq3kV7wuHlTTrPdQfj3IoP6Y)
> 
> A better approach would be to either call Service Catalog ONLY IF there were messages in the queue (check SQS first), or to trigger the Lambda based on events in SQS alone (as mentioned by @wellsiau-aws, prob. better); either of the approaches should take care of the issue

Good info, thank you. and good option would be call Service Catalog API if there are messages in the queue — we are already implementing this approach internally and it works fine. However, triggering the Lambda based on SQS events could lead to issues due to the limitations in Service Catalog's provisioning. Currently, only a maximum of five products can be provisioned in parallel. If more than five account requests are submitted at once, it could create issues with provisioning the sixth account, and there is a risk that the SQS message might be lost. and it might need more changes in the AFT framework to handle it.

**Comment by jpablos11** [comment]
@Tharsan05, thank you for the angle on SQS events and maximum products limitation.

Aside from the method, I think the key conclusion we are reaching is that Service Catalog should only be called if there is something to process, and revising the lambda logic might be a good place to start.

**Comment by jpablos11** [comment]
I just want to provide a quick update, as suspected, the daily consistent cost for Service Catalog went away as soon as I disabled the EventBridge trigger for "aft-account-request-processor".

The trigger was disabled on Nov-13, in the Cost Explorer screenshot below you can see the partial day (day cost went down already), and Nov-14 (yesterday) it completely went away:
![image](https://github.com/user-attachments/assets/ceec7550-31e9-4218-9a78-8246399846c0)

I used heavily AFT Account/Global customizations yesterday through CodePipeline to provision VPCs, etc., but this did not cause any Service Catalog charges, because I believe, it is only used in early stages of ADD/UPDATE Account Request operations.

As we know, Account Request operations (provision accounts, bring under management, update meta parameters) do not happen often, so I think it is not necessary to run the "aft-account-request-processor" lambda every 5 minutes, specially if it is calling Service Catalog every time within.

For the time being, I will keep the EventBridge lambda trigger disabled manually, and only enable it when I have account requests to process, easy enough to do, just an unnecessary manual step to remember so I hope this can be investigated and released as a bug fix/improvement soon.

**Comment by Tharsan05** [comment]
Since we can not directly update the code in class, I have managed fix the issue by adding the folliwing code under line 55 in the `modules/aft-account-request-framework/iam/role-policies/lambda-account-request-processor.tpl` file and AFT works like charm, we have managed to save $350 monthly. 

```
sql_client = boto3.client('sqs')
account_info = get_account_info()
queue_name  = utils.get_ssm_parameter_value(
    aft_management_session, utils.SSM_PARAM_ACCOUNT_REQUEST_QUEUE
    )
    response = sql_client.get_queue_attributes(
        QueueUrl=f'https://sqs.{account_info["region"]}.amazonaws.com/{account_info["account"]}/{queue_name}',
        AttributeNames=['ApproximateNumberOfMessages']
    )

    message_count = int(response['Attributes']['ApproximateNumberOfMessages'])
    if message_count == 0:
        logger.info("SQS is empty. There is no account requests, exiting")
        return None
```

**Comment by harmalh** [comment]
Adding concrete technical details from our deployment, since we hit the same behavior and validated cost impact.

### Problem
`aft-account-request-processor` is scheduled every 5 minutes (`rate(5 minutes)`), and Service Catalog-related logic executes even when there are no pending account requests.

In our AFT setup, this produced a very stable baseline cost:
- ~1152 Service Catalog API calls/day
- ~$0.8064/day (eu-central-1)
- ~$24/month

When idle, these calls are mostly overhead (no account vending activity).

### Goal
Reduce idle Service Catalog calls while preserving existing behavior and concurrency controls by default.

### Technical details (minimal and backward-compatible)
I suggest two incremental changes:

1. **Configurable processor schedule**
- Add root/module variable: `aft_account_request_processor_schedule_expression`
- Default: `rate(5 minutes)` (preserve current behavior)
- Use it in `modules/aft-account-request-framework/eventbridge.tf` for `aws_cloudwatch_event_rule.aft_account_request_processor`.

2. **Optional queue-empty guard before Service Catalog operations**
- Add root/module variable: `aft_skip_service_catalog_when_queue_empty` (bool)
- Default: `false` (preserve current behavior)
- If `true`, check queue state first and exit early when empty, before Service Catalog calls.

### Practical implementation options

**Option A (lowest-risk, no new IAM action): reorder existing flow**
In `src/aft_lambda/aft_account_request_framework/aft_account_request_processor.py`:
- call `receive_sqs_message(...)` first
- if `None`: return
- only then run `associate_aft_service_role_with_account_factory()` and `provisioning_threshold_reached(...)`

This reuses existing `sqs:ReceiveMessage` permission and keeps logic simple.

**Option B (explicit queue depth check)**
- Add `sqs:GetQueueAttributes` permission in `modules/aft-account-request-framework/iam/role-policies/lambda-account-request-processor.tpl`
- Add helper in `sources/aft-lambda-layer/aft_common/sqs.py`
- Check queue depth first and exit if 0

Both options can keep current defaults unchanged.

### Why this matters
For organizations with bursty account vending (not continuous), this avoids paying a fixed monthly idle tax while keeping AFT ready for scale.

---

I understand from `CONTRIBUTING.md` that direct contributions may be limited. If useful, I can still share a PR-quality patch/diff in this issue (file-by-file) so maintainers can apply internally with less effort.

**Comment by yashl** [comment]
Hi all, in the last two releases we've made significant reductions to Service Catalog API call volume during account provisioning, which should resolve the timeout reported here.

**1.19.0** introduced caching for the `provisioned_product_exists` check so it runs once per Lambda invocation instead of being called repeatedly. It also replaced `ScanProvisionedProducts` with `ListEnabledBaselines` for the concurrency threshold check, removing another SC API dependency from the hot path.

**1.20.0** went further and replaced the enrollment check entirely. Previously, it called `GetProvisionedProductOutputs` on every existing provisioned product to match by email, scaling linearly with account count. The new approach uses `Organizations.ListAccounts` to locate the account, then makes a single filtered `SearchProvisionedProducts` call by account ID, reducing SC API volume from ~N individual lookups to one targeted query.

Together these changes should eliminate the Lambda timeouts at scale. We're closing this issue as resolved. If you're still experiencing timeouts after upgrading to 1.20.0, please open a new issue with your account count and the specific Lambda/step that's timing out so we can investigate further.
